xds/internal/resolver/xds_resolver_test.go

```diff
@@ -1178,295 +1265,349 @@ func (s) TestXDSResolverMaxStreamDuration(t *testing.T) {
+			Name: rdsName,
+			VirtualHosts: []*v3routepb.VirtualHost{{
+				Domains: []string{ldsName},
+				Routes: []*v3routepb.Route{{
+					Match: &v3routepb.RouteMatch{PathSpecifier: &v3routepb.RouteMatch_Prefix{Prefix: "/"}},
+					Action: &v3routepb.Route_Route{Route: &v3routepb.RouteAction{
+						ClusterSpecifier: &v3routepb.RouteAction_WeightedClusters{WeightedClusters: &v3routepb.WeightedCluster{
+							Clusters: []*v3routepb.WeightedCluster_ClusterWeight{
+								{
+									Name:   "NEW-CLUSTER",
+									Weight: &wrapperspb.UInt32Value{Value: 100},
+								},
+							},
+						}},
+					}},
+				}},
+			}},
+		}},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
 
-	gotState, err = tcc.stateCh.Receive(ctx)
+	// Read the update pushed by the resolver to the ClientConn and ensure the
+	// old cluster is present in the service config. Also ensure that the newly
+	// returned config selector does not hold a reference to the old cluster.
+	val, err = tcc.stateCh.Receive(ctx)
 	if err != nil {
-		t.Fatalf("Error waiting for UpdateState to be called: %v", err)
+		t.Fatalf("Timeout waiting for an update from the resolver: %v", err)
 	}
-	rState = gotState.(resolver.State)
+	rState = val.(resolver.State)
 	if err := rState.ServiceConfig.Err; err != nil {
-		t.Fatalf("ClientConn.UpdateState received error in service config: %v", rState.ServiceConfig.Err)
+		t.Fatalf("Received error in service config: %v", rState.ServiceConfig.Err)
 	}
-	wantJSON2 := `{"loadBalancingConfig":[{
-	 "xds_cluster_manager_experimental":{
-	   "children":{
-		 "cluster:test-cluster-1":{
-		   "childPolicy":[{"cds_experimental":{"cluster":"test-cluster-1"}}]
-		 },
-		 "cluster:NEW":{
-		   "childPolicy":[{"cds_experimental":{"cluster":"NEW"}}]
-		 }
-	   }
-	 }}]}`
-	wantSCParsed2 := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(wantJSON2)
-	if !internal.EqualServiceConfigForTesting(rState.ServiceConfig.Config, wantSCParsed2.Config) {
-		t.Errorf("ClientConn.UpdateState received different service config")
+	wantSCParsed = internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(`
+{
+	"loadBalancingConfig": [
+		{
+		  "xds_cluster_manager_experimental": {
+			"children": {
+			  "cluster:OLD-CLUSTER": {
+				"childPolicy": [
+				  {
+					"cds_experimental": {
+					  "cluster": "OLD-CLUSTER"
+					} 
+				  } 
+				] 
+			  }, 
+			  "cluster:NEW-CLUSTER": {
+				"childPolicy": [
+				  {
+					"cds_experimental": {
+					  "cluster": "NEW-CLUSTER"
+					} 
+				  } 
+				] 
+			  } 
+			} 
+		  } 
+		} 
+	  ] 
+}`)
+	if !internal.EqualServiceConfigForTesting(rState.ServiceConfig.Config, wantSCParsed.Config) {
+		t.Errorf("Received unexpected service config")
 		t.Error("got: ", cmp.Diff(nil, rState.ServiceConfig.Config))
-		t.Fatal("want: ", cmp.Diff(nil, wantSCParsed2.Config))
+		t.Fatal("want: ", cmp.Diff(nil, wantSCParsed.Config))
```

Comment:
Same comment here abut combining. This is the first time I've seen three operations for this unexpected got want deal, instead of one operation.